src/cunumeric/stat/histogram.cc

```diff
@@ -0,0 +1,183 @@
+#ifdef _DEBUG
+#include <iostream>
+#include <iterator>
+#endif
+
+namespace cunumeric {
+using namespace legate;
+
+namespace detail {
+
+// RO accessor (size, pointer) extractor:
+//
+template <typename VAL>
+std::tuple<size_t, const VAL*> get_accessor_ptr(const AccessorRO<VAL, 1>& src_acc,
+                                                const Rect<1>& src_rect)
+{
+  size_t src_strides[1];
+  const VAL* src_ptr = src_acc.ptr(src_rect, src_strides);
+  assert(src_strides[0] == 1);
+  //
+  // const VAL* src_ptr: need to create a copy with create_buffer(...);
+  // since src will get sorted (in-place);
+  //
+  size_t src_size = src_rect.hi - src_rect.lo + 1;
+  return std::make_tuple(src_size, src_ptr);
+}
+// RD accessor (size, pointer) extractor:
+//
+template <typename VAL>
+std::tuple<size_t, VAL*> get_accessor_ptr(const AccessorRD<SumReduction<VAL>, true, 1>& src_acc,
+                                          const Rect<1>& src_rect)
+{
+  size_t src_strides[1];
+  VAL* src_ptr = src_acc.ptr(src_rect, src_strides);
+  assert(src_strides[0] == 1);
+  //
+  // const VAL* src_ptr: need to create a copy with create_buffer(...);
+  // since src will get sorted (in-place);
+  //
+  size_t src_size = src_rect.hi - src_rect.lo + 1;
+  return std::make_tuple(src_size, src_ptr);
+}
+// accessor copy utility:
+//
+template <typename VAL>
+std::tuple<size_t, Buffer<VAL>, const VAL*> make_accessor_copy(const AccessorRO<VAL, 1>& src_acc,
+                                                               const Rect<1>& src_rect)
+{
+  size_t src_strides[1];
+  const VAL* src_ptr = src_acc.ptr(src_rect, src_strides);
+  assert(src_strides[0] == 1);
+  //
+  // const VAL* src_ptr: need to create a copy with create_buffer(...);
+  // since src will get sorted (in-place);
+  //
+  size_t src_size      = src_rect.hi - src_rect.lo + 1;
+  Buffer<VAL> src_copy = create_buffer<VAL>(src_size);
+  return std::make_tuple(src_size, src_copy, src_ptr);
+}
+}  // namespace detail
+
+template <Type::Code CODE>
+struct HistogramImplBody<VariantKind::CPU, CODE> {
+  using VAL = legate_type_of<CODE>;
+
+  // for now, it has been decided to hardcode these types:
+  //
+  using BinType    = double;
+  using WeightType = double;
+
+  // in the future we might relax relax that requirement,
+  // but complicate dispatching:
+  //
+  // template <typename BinType = VAL, typename WeightType = VAL>
+  void operator()(const AccessorRO<VAL, 1>& src,
+                  const Rect<1>& src_rect,
+                  const AccessorRO<BinType, 1>& bins,
+                  const Rect<1>& bins_rect,
+                  const AccessorRO<WeightType, 1>& weights,
+                  const Rect<1>& weights_rect,
+                  const AccessorRD<SumReduction<WeightType>, true, 1>& result,
+                  const Rect<1>& result_rect) const
+  {
+    auto&& [src_size, src_copy, src_ptr] = detail::make_accessor_copy(src, src_rect);
+    std::copy_n(src_ptr, src_size, src_copy.ptr(0));
```

Comment:
This is CPU code. I don't see how thrust could accelerate that. In fact, I suspect host version of `thrust::copy` just calls `std::copy`. One reason I actually do the copy outside of `make_accessor_copy()` is b/c I want to avoid passing to it additional arguments (like the exec-policy and / or stream, etc.). 

The semantics of `make_accessor_copy()` was just to make the copy, not fill it (what if I want to fill it w/ something other than content of `src_ptr`? Or. what if I want to use a different exec_policy for debugging purposes.

Reply:
> This is CPU code. I don't see how thrust could accelerate that. In fact, I suspect host version of thrust::copy just calls std::copy.

That's true, I should have put this note on the OpenMP file. 

In the CPU case `thrust::copy` is indeed probably equivalent to `std::copy`. However, in the OpenMP file `thrust::copy` will be using multiple threads, whereas `std::copy` will always use a single thread. And if we switch to `thrust::copy` there then we might as well use it here too, so we retain a uniform implementation.

Reply:
Fixed according to the 4 points we agreed:

1. unify accessor helpers to one undifferentiated form for every ecosystem (GPU/CPU/OMP);
2. put them in a header (*_gen.h), under a deeper nested namespace, say, cunumeric::detail::accessor namespace;
3. move the copy inside make_accessor_copy();
4. do the copying via thrust::copy_n();